File: Mobile/remetask/lib/Models/Task.dart
import 'package:json_annotation/json_annotation.dart';

import 'TaskGroup.dart';
part 'Task.g.dart';

@JsonSerializable(explicitToJson: true)
class Task {
  Task(this.title, this.description, this.deadline, this.isCompleted, this.completionDate, this.priority, this.taskGroup, this.taskGroupId);

  int? id;
  String title;
  String description;
  DateTime deadline;
  bool? isCompleted;
  DateTime? completionDate;
  int priority;
  @JsonKey(toJson: null, includeIfNull: false)
  TaskGroup? taskGroup;
  int? taskGroupId;

  factory Task.fromJson(Map<String, dynamic> json) => _$TaskFromJson(json);

  Map<String, dynamic> toJson() => _$TaskToJson(this);

  bool isDeadline({int deadlineDaysThreshold = 5})
  {
    var timeLeft =  deadline.difference(DateTime.now());
    if(timeLeft.inDays <= deadlineDaysThreshold && !isCompleted!)
      return true;
    return false;
  }
}

File: API/RemeTask/RemeTask/Models/Task.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Linq;
using System.Threading.Tasks;

namespace RemeTask.Models
{
    public class Task : Entity
    {
        public string Title { get; set; }
        public string Description { get; set; }
        public DateTime Deadline { get; set; }
        public bool IsCompleted { get; set; }
        public DateTime? CompletionDate { get; set; }
        public int Priority { get; set; }

        public int TaskGroupId { get; set; }
        public TaskGroup TaskGroup { get; set; }
        public ICollection<TaskNote> TaskNotes { get; set; }
    }
}
